gitfile/tokenfetcher_spi.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2022 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitfile
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/mshaposhnik/service-provider-integration-scm-file-retriever/gitfile/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SpiTokenFetcher token fetcher implementation that looks for token in the specific ENV variable.
+type SpiTokenFetcher struct {
+	k8sClient client.Client
+	namespace string
+}
+
+const (
+	namespaceFile = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+	letterBytes   = "abcdefghijklmnopqrstuvwxyz1234567890"
+)
+
+func NewSpiTokenFetcher() *SpiTokenFetcher {
+	namespace, err := ioutil.ReadFile(namespaceFile)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	scheme := runtime.NewScheme()
+	if err = corev1.AddToScheme(scheme); err != nil {
+		panic(err.Error())
+	}
+
+	if err = v1beta1.AddToScheme(scheme); err != nil {
+		panic(err.Error())
+	}
+
+	// creates the client
+	k8sClient, err := client.New(config, client.Options{Scheme: scheme})
+	if err != nil {
+		panic(err.Error())
+	}
+	return &SpiTokenFetcher{k8sClient: k8sClient, namespace: string(namespace)}
+}
+
+func (s *SpiTokenFetcher) BuildHeader(ctx context.Context, repoUrl string, loginCallback func(url string)) (*HeaderStruct, error) {
+
+	var tBindingName = "file-retriever-binging-" + randStringBytes(6)
+	var secretName = "file-retriever-secret-" + randStringBytes(5)
+
+	// create binding
+	newBinding := newSPIATB(tBindingName, s, repoUrl, secretName)
+	err := s.k8sClient.Create(ctx, newBinding)
+	if err != nil {
+		zap.L().Error("Error creating Token Binding item:", zap.Error(err))
+		return nil, err
+	}
+
+	// scheduling the binding cleanup
+	defer func() {
+		// clean up token binding
+		err = s.k8sClient.Delete(ctx, newBinding)
+		if err != nil {
+		}
+	}()
+
+	// now re-reading SPITokenBinding to get updated fields
+	var tokenName string
+	for {
```

Comment:
Run this asynchronously with some timeout?

Reply:
tiemout - yes, asynchronous - doesnt make sense, we have nothing to do in parallel until this routine is completed.